src/internal/packager2/pull.go

```diff
@@ -0,0 +1,234 @@
+	tmpDir, err := utils.MakeTempDir(config.CommonOptions.TempDirectory)
+	if err != nil {
+		return err
+	}
+	defer os.Remove(tmpDir)
+	if shasum != "" {
+		src = fmt.Sprintf("%s@sha256:%s", src, shasum)
+	}
+	arch := config.GetArch()
+	remote, err := zoci.NewRemote(src, oci.PlatformForArch(arch))
+	if err != nil {
+		return err
+	}
+	desc, err := remote.ResolveRoot(ctx)
+	if err != nil {
+		return fmt.Errorf("could not fetch images index: %w", err)
+	}
+	layersToPull := []ocispec.Descriptor{}
+	if supportsFiltering(desc.Platform) {
+		pkg, err := remote.FetchZarfYAML(ctx)
+		if err != nil {
+			return err
+		}
+		pkg.Components, err = filter.Apply(pkg)
+		if err != nil {
+			return err
+		}
+		layersToPull, err = remote.LayersFromRequestedComponents(ctx, pkg.Components)
+		if err != nil {
+			return err
+		}
+	}
+	_, err = remote.PullPackage(ctx, tmpDir, config.CommonOptions.OCIConcurrency, layersToPull...)
+	if err != nil {
+		return err
+	}
+	allTheLayers, err := filepath.Glob(filepath.Join(tmpDir, "*"))
+	if err != nil {
+		return err
+	}
+	err = archiver.Archive(allTheLayers, tarPath)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func pullHTTP(ctx context.Context, src, tarPath, shasum string) error {
+	if !config.CommonOptions.Insecure && shasum == "" {
+		return errors.New("remote package provided without shasum while insecure is not enabled")
+	}
+	f, err := os.Create(tarPath)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, src, nil)
+	if err != nil {
+		return err
+	}
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != http.StatusOK {
+		_, err := io.Copy(io.Discard, resp.Body)
+		if err != nil {
+			return err
+		}
+		return fmt.Errorf("unexpected http response status code %s for source %s", resp.Status, src)
+	}
+	_, err = io.Copy(f, resp.Body)
+	if err != nil {
+		return err
+	}
+	// Check checksum if src included one.
+	if shasum != "" {
+		received, err := helpers.GetSHA256OfFile(tarPath)
+		if err != nil {
+			return err
+		}
+		if received != shasum {
+			return fmt.Errorf("shasum mismatch for file %s, expected %s bu got %s ", tarPath, shasum, received)
+		}
+	}
+	return nil
+}
+
+func nameFromMetadata(path string) (string, error) {
+	var pkg v1alpha1.ZarfPackage
+	err := archiver.Walk(path, func(f archiver.File) error {
+		if f.Name() == layout.ZarfYAML {
+			b, err := io.ReadAll(f)
+			if err != nil {
+				return err
+			}
+			if err := goyaml.Unmarshal(b, &pkg); err != nil {
+				return err
+			}
+		}
+		return nil
+	})
+	if err != nil {
+		return "", err
+	}
+	if pkg.Metadata.Name == "" {
+		return "", fmt.Errorf("%s does not contain a zarf.yaml", path)
+	}
+
+	arch := config.GetArch(pkg.Metadata.Architecture, pkg.Build.Architecture)
+	if pkg.Build.Architecture == zoci.SkeletonArch {
```

Comment:
Should we validate that theses are not different instead of making them the same?

Reply:
This code came from the old code. After reading it I am a bit confused of the purpose? Should the architecture not be whatever the `pkg.Build.Architecture` states?

Reply:
The `config.GetArch` function returns the highest priority. First it gets the global cli arg `--architecture` then whichever parameters are called in first, then it gets the default runtime. In this case we call `config.GetArch` with no args in `pullOCI` then we call it again here. 

Given this I think we should instead just use pkg.Build.Architecture without `config.GetArch` here, as we will always want what is actually in the package.